src/cpp/session/modules/shiny/SessionShiny.cpp

```diff
@@ -196,6 +197,87 @@ std::string getLastFunction(const std::string& fileContents)
    return function;
 }
 
+const char * const kShinyAppTypeSingleFile = "type_single_file";
+const char * const kShinyAppTypeMultiFile =  "type_multi_file";
+
+FilePath shinyTemplatePath(const std::string& name)
+{
+   return session::options().rResourcesPath().childPath("templates/shiny/" + name);
+}
+
+Error copyTemplateFile(const std::string& templateFileName,
+                      const FilePath& target,
+                      json::Object* pResult)
+{
+   FilePath templatePath = shinyTemplatePath(templateFileName);
+   Error error = templatePath.copy(target);
+   if (error)
+      LOG_ERROR(error);
+   
+   std::string aliasedPath = module_context::createAliasedPath(target);
+   (*pResult)[aliasedPath] = (error == Success());
+   return error;
+}
+
+Error createShinyApp(const json::JsonRpcRequest& request,
+                     json::JsonRpcResponse* pResponse)
+{
+   json::Object result;
+   
+   std::string appName;
+   std::string appType;
+   std::string appDirString;
+   
+   Error error = json::readParams(request.params,
+                                  &appName,
+                                  &appType,
+                                  &appDirString);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return error;
+   }
+   
+   FilePath appDir = module_context::resolveAliasedPath(appDirString);
+   
+   // collect the files we want to generate
+   std::vector<std::string> templateFiles;
+   if (appType == kShinyAppTypeSingleFile)
+   {
+      templateFiles.push_back("app.R");
+   }
+   else if (appType == kShinyAppTypeMultiFile)
+   {
+      templateFiles.push_back("ui.R");
+      templateFiles.push_back("server.R");
+   }
+   
+   // by default, report failure to create a file
+   BOOST_FOREACH(const std::string& fileName, templateFiles)
```

Comment:
For error reporting here you may want to use our ClientInfo sideband mechanism:
https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/server/remote/RemoteServerError.java#L78-L82

When setting an error you can also set option client info: https://github.com/rstudio/rstudio/blob/master/src/cpp/core/include/core/json/JsonRpc.hpp#L1028

By using this you can return a more descriptive error indicating e.g. a file exists, an error occurred while copying, etc.